interfaces/policy/policy.go

```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
 
 // ConnectCandidate represents a candidate connection.
 type ConnectCandidate struct {
-	// TODO: later we need to carry dynamic attributes once we have those
-	Plug                *snap.PlugInfo
+	Plug *interfaces.ConnectedPlug
+	// Static + dynamic attributes, merged lazily by plugAttrs below (FIXME: remove it)
+	PlugAttrs           map[string]interface{}
 	PlugSnapDeclaration *asserts.SnapDeclaration
 
-	Slot                *snap.SlotInfo
+	Slot *interfaces.ConnectedSlot
+	// Static + dynamic attributes, merged lazily by slotAttrs below (FIXME: remove it)
+	SlotAttrs           map[string]interface{}
 	SlotSnapDeclaration *asserts.SnapDeclaration
 
 	BaseDeclaration *asserts.BaseDeclaration
 }
 
+func mergedAttributes(staticAttrs, dynamicAttrs map[string]interface{}) map[string]interface{} {
+	merged := make(map[string]interface{})
+	for k, v := range staticAttrs {
+		merged[k] = v
+	}
+	for k, v := range dynamicAttrs {
+		if _, ok := merged[k]; ok {
+			// Safeguard. This should never happen as it's prevented
+			// when attributes are populated at higher levels.
+			panic(fmt.Sprintf("internal error: attempted to overwrite static attribute %q", k))
+		}
+		merged[k] = v
+	}
+	return merged
+}
+
 func (connc *ConnectCandidate) plugAttrs() map[string]interface{} {
-	return connc.Plug.Attrs
+	// FIXME: change policy code to use Attrer interface, remove merging.
+	if connc.PlugAttrs == nil {
+		connc.PlugAttrs = mergedAttributes(connc.Plug.StaticAttrs(), connc.Plug.DynamicAttrs())
```

Comment:
do we have tests for when DynamicAttrs is not empty?

I'm not spotting them in policy_test.go, but I might be missing them

Reply:
You're right, they were missing. Added.